types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,658 @@
+	}
+
+	wl_list_remove(&connector->link);
+	wl_list_remove(&connector->destroy.link);
+	free(connector);
+}
+
+static void drm_lease_device_v1_destroy(
+		struct wlr_drm_lease_device_v1 *device) {
+	if (!device) {
+		return;
+	}
+
+	struct wlr_drm_backend *backend =
+			get_drm_backend_from_backend(device->backend);
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_device_v1 for %s",
+			backend->name);
+
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_resource_set_user_data(resource, NULL);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->requests) {
+		struct wlr_drm_lease_request_v1 *request =
+				drm_lease_request_v1_from_resource(resource);
+		drm_lease_request_v1_destroy(request);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->leases) {
+		struct wlr_drm_lease_v1 *lease = drm_lease_v1_from_resource(resource);
+		drm_lease_v1_destroy(lease);
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector, *tmp_connector;
+	wl_list_for_each_safe(connector, tmp_connector, &device->connectors, link) {
+		drm_lease_connector_v1_destroy(connector);
+	}
+
+	wl_list_remove(&device->link);
+	wlr_global_destroy_safe(device->global, device->manager->display);
+
+	free(device);
+}
+
+struct wlr_drm_lease_v1 *wlr_drm_lease_request_v1_grant(
+		struct wlr_drm_lease_request_v1 *request) {
+	assert(request->lease);
+
+	wlr_log(WLR_DEBUG, "Attempting to grant request %p", request);
+
+	struct wlr_drm_lease_v1 *lease = request->lease;
+	if (request->invalid) {
+		wlr_log(WLR_ERROR, "Invalid lease request");
+		wp_drm_lease_v1_send_finished(lease->resource);
+		return NULL;
+	}
+
+	/* Transform connectors list into wlr_output for leasing */
+	struct wlr_output *outputs[request->n_connectors + 1];
+	for(size_t i = 0; i < request->n_connectors; ++i) {
+		outputs[i] = request->connectors[i]->output;
+	}
+
+	int fd = wlr_drm_create_lease(outputs, request->n_connectors,
+			&lease->lessee_id);
+	if (fd < 0) {
+		wlr_log_errno(WLR_ERROR, "drm_create_lease failed");
+		return NULL;
+	}
+
+	lease->connectors = calloc(request->n_connectors,
+			sizeof(struct wlr_drm_lease_connector_v1));
+	lease->n_connectors = request->n_connectors;
+	for (size_t i = 0; i < request->n_connectors; ++i) {
+		lease->connectors[i] = request->connectors[i];
+		lease->connectors[i]->active_lease = lease;
+	}
+
+	wlr_log(WLR_DEBUG, "Granting request %p", request);
+
+	wp_drm_lease_v1_send_lease_fd(lease->resource, fd);
+	close(fd);
+
+	return lease;
+}
+
+void wlr_drm_lease_request_v1_reject(
+		struct wlr_drm_lease_request_v1 *request) {
+	assert(request && request->lease);
+
+	wlr_log(WLR_DEBUG, "Rejecting request %p", request);
+
+	request->invalid = true;
+	wp_drm_lease_v1_send_finished(request->lease->resource);
+}
+
+void wlr_drm_lease_v1_revoke(struct wlr_drm_lease_v1 *lease) {
```

Comment:
It would be nice if `wlr_drm_lease_v1_revoke` also destroyed the lease. That way we wouldn't rely on clients behaving nicely and destroying the resource when they get a `finished` event. In other words, a memory leak in the client doesn't trigger a memory leak in the compositor.